adapters/repos/db/vector/cuvs/index.go

```diff
@@ -0,0 +1,895 @@
+	}
+
+	_, err := index.dlpackTensor.ToHost(index.cuvsResource)
+	slice, err := index.dlpackTensor.Slice()
+	if err != nil {
+		return err
+	}
+
+	newSlice := [][]float32{}
+
+	newIndex := uint32(0)
+
+	ids_cuvsIds := make([]uint32, len(ids))
+
+	for i := range ids {
+		var exists bool
+		ids_cuvsIds[i], exists = index.idCuvsIdMap.GetCuvsId(ids[i])
+		if !exists {
+			return errors.Errorf("idCuvsIdMap does not contain id %d", ids[i])
+		}
+	}
+
+	for i := range slice {
+		if !slices.Contains(index.cuvsFilter, uint32(i)) && !slices.Contains(ids_cuvsIds, uint32(i)) {
+			newSlice = append(newSlice, slice[i])
+
+			// remap
+			WeaviateId, _ := index.idCuvsIdMap.GetWeaviateId(uint32(i))
+
+			index.idCuvsIdMap.DeleteByCuvsId(uint32(i))
+
+			index.idCuvsIdMap.Insert(newIndex, WeaviateId)
+			newIndex++
+		} else {
+			index.idCuvsIdMap.DeleteByCuvsId(uint32(i))
+		}
+	}
+
+	index.cuvsNumFiltered = 0
+	index.cuvsFilter = []uint32{}
+
+	index.dlpackTensor.Close()
+
+	index.count = uint64(len(newSlice))
+	index.nextCuvsId = uint32(len(newSlice))
+
+	newTensor, err := cuvs.NewTensor(newSlice)
+	if err != nil {
+		return err
+	}
+
+	_, err = newTensor.ToDevice(index.cuvsResource)
+	if err != nil {
+		return err
+	}
+
+	index.dlpackTensor = &newTensor
+
+	err = cagra.BuildIndex(*index.cuvsResource, index.cuvsIndexParams, index.dlpackTensor, index.cuvsIndex)
+	if err != nil {
+		return err
+	}
+	println("done delete with rebuild")
+
+	return nil
+}
+
+func (index *cuvs_index) createAllowList() []uint32 {
+	list := []uint32{}
+
+	for i := 0; i < int(index.nextCuvsId); i++ {
+		if !slices.Contains(index.cuvsFilter, uint32(i)) {
+			list = append(list, uint32(i))
+		}
+	}
+
+	return list
+}
+
+func (index *cuvs_index) SearchByVector(ctx context.Context, vector []float32, k int, allow helpers.AllowList) ([]uint64, []float32, error) {
+	index.Lock()
+	defer index.Unlock()
+	queries, err := cuvs.NewTensor([][]float32{vector})
+	if err != nil {
+		return nil, nil, err
+	}
+	neighbors, err := cuvs.NewTensorOnDevice[uint32](index.cuvsResource, []int64{int64(1), int64(k)})
+	if err != nil {
+		return nil, nil, err
+	}
+	distances, err := cuvs.NewTensorOnDevice[float32](index.cuvsResource, []int64{int64(1), int64(k)})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	_, err = queries.ToDevice(index.cuvsResource)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cagra.SearchIndex(*index.cuvsResource, index.cuvsSearchParams, index.cuvsIndex, &queries, &neighbors, &distances, index.createAllowList())
+
+	_, err = neighbors.ToHost(index.cuvsResource)
+	if err != nil {
+		return nil, nil, err
+	}
+	_, err = distances.ToHost(index.cuvsResource)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	neighborsSlice, err := neighbors.Slice()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	distancesSlice, err := distances.Slice()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	neighborsResultSlice := make([]uint64, k)
+
+	for i := range neighborsSlice[0] {
```

Comment:
Is this for debugging? Otherwise we should check and throw error if `neighborsSlice` has `0`